cocotb/__init__.py

```diff
@@ -106,8 +107,36 @@ def _reopen_stream_with_buffering(stream_name):
 plusargs = {}
 """A dictionary of "plusargs" handed to the simulation."""
 
-# To save typing provide an alias to scheduler.add
-fork = scheduler.add
+
+def current_test() -> cocotb.decorators.RunningTest:
+    """Returns a reference to the currently running test"""
+    global scheduler
+    return scheduler._test
+
+
+def create_task(coro: Union[Coroutine, cocotb.decorators.RunningTask]) -> cocotb.decorators.RunningTask:
+    """
+    """
+    global scheduler
+    return scheduler.add(coro)
+
+
+def fork(coro: Union[Coroutine, cocotb.decorators.RunningTask]):
+    global scheduler
+
+    co = create_task(coro)
+
+    async def watcher():
+        try:
+            await co
+        except Exception as e:
+            if cocotb.triggers.Join(co) not in scheduler._trigger2coros:
+                current_test().abort(e)
+
+    create_task(watcher())
+
+    return co
```

Comment:
At first glance, this looks quite tidy - thanks!

Reply:
But, it is also incorrect. You might have missed my commit message.

https://github.com/cocotb/cocotb/blob/d1a305da30fa0ffdd0174358657fa94ff778bcb6/cocotb/scheduler.py#L401-L404

I can't check for the `Join` trigger because the first thing it does before servicing all the waiting coroutines is pop the trigger. I'm not sure an easy way around this... I feel like instead of popping the trigger to prevent a potential infinite recursion, the entire `triggers2coros` object should be immutably copied and cleared for the next "wave" of schedulings.

Any help would be appreciated.

Reply:
Newest commit "fixes" this issue. But there are still yet more issues which I guess hasn't been seen before? When there are multiple coroutines (B and C) waiting on coroutine A, if coroutine A ends with an exception, it will rethrow in one coroutine (B) **and add B's stacktrace to the stacktrace object**. So when coroutine C handles the exception, the stacktrace will include B's stack...

Reply:
Well I used `remove_traceback_frame` and we are good, but this is still an outstanding issue in general.